anonymous_helpline_chatbot/core/invitations.py

```diff
@@ -1,89 +1,131 @@
+                an interesting condition: the user row matches an invitation row if the user is the invitation operator
+                AND the invitation client is the client we're currently processing. This way we get rid of other
+                invitations sent for this operator.
+                After that we only keep the operators which don't yet have an invitation sent to the client
+                (`WHERE ... AND sent_invitations.client_chat_id IS NULL`)
+
+            That's it! Now we have the list of operators which should receive an invitation to the client.
+            """
+            # FIXME: fuck, this query relies on tables `users`, `conversations`, and `sent_invitations`, while
+            #  `InvitationsController` should only rely on `sent_invitations`. Have to do something with it. Either
+            #  somehow improve the API of the three controllers, or document, that the controllers can't actually be
+            #  safely replaced with other classes of the same interface
+            cursor.execute("SELECT users.chat_id "
+                           "FROM users "
+
+                           "   LEFT OUTER JOIN conversations ON users.chat_id = conversations.operator_chat_id "
+                           "                                    OR users.chat_id = conversations.client_chat_id "
+
+                           "   LEFT OUTER JOIN sent_invitations ON users.chat_id = sent_invitations.operator_chat_id "
+                           "                                       AND sent_invitations.client_chat_id = %s "
+
+                           "WHERE users.is_operator "
+                           "  AND users.chat_id != %s"
+                           "  AND conversations.operator_chat_id IS NULL "
+                           "  AND sent_invitations.client_chat_id IS NULL ",
+                           (client_chat_id, client_chat_id))
+
+            for operator_chat_id, in cursor.fetchall():
+                self._invite_operator_to_client(cursor, operator_chat_id, client_chat_id)
+
+    def invite_for_operator(self, operator_chat_id: int) -> None:
+        with self._conn_pool.PrettyCursor() as cursor,\
+                self.conversations_controller.get_conversations_requesters_with_plocking() as conversations_requesters:
+
+            # Prevent any invitations from being sent or deleted by parallel transactions until this one completes,
+            # because, for example, a parallel transaction might try to delete an invitation which we are going to send,
+            # but have not sent yet
+            cursor.execute("LOCK TABLE sent_invitations IN SHARE MODE")
+
+            for client_chat_id in conversations_requesters:
+                self._invite_operator_to_client(cursor, operator_chat_id, client_chat_id)
+
+    def clear_invitations_to_client(self, client_chat_id: int) -> bool:
         """
         Remove messages with invitations to a conversation with the client sent to operators
```

Comment:
There should be a comma, but I think this sentence looks even better without the tail at all:
```suggestion
        Remove messages with invitations to a conversation with the client
```
or even
```suggestion
        Remove messages, inviting to a conversation with the client
```